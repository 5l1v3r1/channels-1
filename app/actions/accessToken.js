// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.REQUEST_ACCESS_TOKEN
  };
}

export function receiveAccessToken(json: any): Action {
  return {
    type: actionTypes.RECEIVE_ACCESS_TOKEN,
    payload: json.access_token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.ACCESS_TOKEN_ERROR
  };
}
