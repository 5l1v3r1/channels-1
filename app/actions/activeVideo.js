// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
