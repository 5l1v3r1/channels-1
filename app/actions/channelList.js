// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
