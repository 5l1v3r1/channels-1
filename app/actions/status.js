// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function toggleChannelList(): Action {
  return {
    type: actionTypes.TOGGLE_CHANNEL_LIST
  };
}

export function updateStatus(title: string, thumbnail: string): Action {
  return {
    type: actionTypes.UPDATE_STATUS,
    payload: {
      title,
      thumbnail
    }
  };
}
