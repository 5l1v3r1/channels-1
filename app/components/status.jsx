// @flow
import React from 'react';
import { IconArrowDropDown } from './icons';
import styles from './status.scss';

const Status = ({
  thumbnail,
  title,
  toggleChannelList
}: {
  thumbnail: string,
  title: string,
  toggleChannelList: () => void
}) => (
  <div className={styles.status} onClick={toggleChannelList}>
    <img className={styles.thumbnail} src={thumbnail} />
    <h1 className={styles.title}>{title}</h1>
    <IconArrowDropDown styles={styles} />
  </div>
);

export default Status;
