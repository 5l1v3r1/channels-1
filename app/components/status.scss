@import '../app.global.scss';

.status {
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 11px;
  background-color: $red;
}

.thumbnail {
  width: 45px;
  margin-right: 11px;
  border: 2px solid $white;
  border-radius: 50%;
  box-shadow: 0 1px 3px 1px rgba(46, 46, 46, 0.3);
}

.title {
  width: 100px;
  color: $white;
  font-weight: 300;

  @include font-size(24);
}

.css-truncate-target {
  max-width: 100px !important;
}

.iconArrowDropDown {
  width: 14px;
  margin-left: 0.5rem;
  fill: $white;
}
