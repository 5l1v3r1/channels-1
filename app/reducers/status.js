// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, StatusState } from '../constants/typeAliases';
import placeholder from '../images/placeholder.jpg';

const initialState = {
  thumbnail: placeholder,
  title: 'Channels'
};

export default function status(state: StatusState = initialState, { type, payload }: Action): StatusState {
  switch (type) {
    case actionTypes.UPDATE_STATUS:
      return {
        title: payload.title,
        thumbnail: payload.thumbnail
      };
    default:
      return state;
  }
}
