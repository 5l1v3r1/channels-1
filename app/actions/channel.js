// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import ChannelType from '../dataTypes/channelType';
import { toggleChannelList, updateStatus } from './status';
import { enableSearchBar } from './searchBar';

type ThunkAction = (dispatch: Dispatch) => any;

export function selectChannel(channel: ChannelType): ThunkAction {
  return (dispatch: Dispatch) => {
    const { id, title, thumbnail } = channel;
    dispatch(toggleChannelList());
    dispatch(updateActiveChannel(id));
    dispatch(updateStatus(title, thumbnail));
    dispatch(enableSearchBar(title));
  };
}

export function updateActiveChannel(id: string): Action {
  return {
    type: actionTypes.UPDATE_ACTIVE_CHANNEL,
    payload: id
  };
}
