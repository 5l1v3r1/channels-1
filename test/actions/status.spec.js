import * as statusActions from '../../app/actions/status';

describe('status actions', () => {
  it('should create an action to toggle the channel list', () => {
    expect(statusActions.toggleChannelList()).toMatchSnapshot();
  });

  it('should create an action to update status', () => {
    const title = 'DevTips';
    const thumbnail = 'https://yt3.ggpht.com/photo.jpg';
    expect(statusActions.updateStatus(title, thumbnail)).toMatchSnapshot();
  });
});
