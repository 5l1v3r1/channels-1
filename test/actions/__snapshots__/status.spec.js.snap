// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`status actions should create an action to toggle the channel list 1`] = `
Object {
  "type": "TOGGLE_CHANNEL_LIST",
}
`;

exports[`status actions should create an action to update status 1`] = `
Object {
  "payload": Object {
    "thumbnail": "https://yt3.ggpht.com/photo.jpg",
    "title": "DevTips",
  },
  "type": "UPDATE_STATUS",
}
`;
